collector/logind_linux.go

```diff
@@ -0,0 +1,241 @@
+// Copyright 2016 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !nologind
+
+package collector
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+
+	"github.com/godbus/dbus"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	logindSubsystem = "logind"
+	dbusObject      = "org.freedesktop.login1"
+	dbusPath        = "/org/freedesktop/login1"
+)
+
+var (
+	// Taken from logind as of systemd v229
+	attrRemoteValues = []string{"true", "false"}
+	attrTypeValues   = []string{"unspecified", "tty", "x11", "wayland", "mir", "web"}
+	attrClassValues  = []string{"user", "greeter", "lock-screen", "background"}
+)
+
+type logindCollector struct {
+	sessionsDesc *prometheus.Desc
+}
+
+type logindDbus struct {
+	conn   *dbus.Conn
+	object dbus.BusObject
+}
+
+type logindSession struct {
+	seat        string
+	remote      string
+	sessionType string
+	class       string
+}
+
+// Struct elements must be public for the reflection magic of godbus to work
+type logindSessionEntry struct {
+	SessionId         string
+	UserId            uint32
+	UserName          string
+	SeatId            string
+	SessionObjectPath dbus.ObjectPath
+}
+
+type logindSeatEntry struct {
+	SeatId         string
+	SeatObjectPath dbus.ObjectPath
+}
+
+func init() {
+	Factories["logind"] = NewLogindCollector
+}
+
+// Takes a prometheus registry and returns a new Collector exposing
+// logind statistics.
+func NewLogindCollector() (Collector, error) {
+	sessionsDesc := prometheus.NewDesc(
+		prometheus.BuildFQName(Namespace, logindSubsystem, "sessions"),
+		"Number of sessions registered in logind.", []string{"seat", "remote", "type", "class"}, nil,
+	)
+
+	return &logindCollector{
+		sessionsDesc: sessionsDesc,
+	}, nil
+}
+
+func (lc *logindCollector) Update(ch chan<- prometheus.Metric) error {
+	c, err := newDbus()
+	if err != nil {
+		return fmt.Errorf("unable to connect to dbus: %s", err)
+	}
+	defer c.conn.Close()
+
+	seats, err := listSeats(c)
+	if err != nil {
+		return fmt.Errorf("unable to get seats: %s", err)
+	}
+
+	sessionList, err := listSessions(c)
+	if err != nil {
+		return fmt.Errorf("unable to get sessions: %s", err)
+	}
+
+	sessions := make(map[logindSession]float64)
+
+	for _, s := range sessionList {
+		session := getSession(c, s)
+		if session != nil {
+			sessions[*session]++
+		}
+	}
+
+	for _, remote := range attrRemoteValues {
```

Comment:
Values outside these are ignored. My idea was to keep the set of exposed metrics as stable as possible.